File: javascript/packages/core/src/models/device.js
/**
 * Device brought in for repair.
 *
 * The only model written the old way: a constructor function plus methods
 * hung on `Device.prototype`, with `Laptop` inheriting through
 * `Object.create(Device.prototype)`. No `class` keyword appears in this file,
 * so every method here is a prototype property, not a class member.
 */

/**
 * @param {number} id
 * @param {number} customerId
 * @param {string} brand
 * @param {string} model
 * @param {string} [serial]
 */
export function Device(id, customerId, brand, model, serial) {
    this.id = id;
    this.customerId = customerId;
    this.brand = brand;
    this.model = model;
    this.serial = serial;
}

/**
 * Human label such as `Framework 13 (SER-0001)`.
 *
 * @returns {string}
 */
Device.prototype.label = function label() {
    return this.serial === undefined
        ? `${this.brand} ${this.model}`
        : `${this.brand} ${this.model} (${this.serial})`;
};

/**
 * Devices without a serial cannot be warranty-claimed.
 *
 * @returns {boolean}
 */
Device.prototype.isWarrantyEligible = function isWarrantyEligible() {
    return this.serial !== undefined;
};

/**
 * Laptop subtype.
 *
 * Inheritance is wired by hand: the parent constructor runs through `.call`
 * with a rebound `this`, and the prototype chain is built with `Object.create`.
 *
 * @param {number} id
 * @param {number} customerId
 * @param {string} brand
 * @param {string} model
 * @param {string | undefined} serial
 * @param {number} screenInches
 */
export function Laptop(id, customerId, brand, model, serial, screenInches) {
    Device.call(this, id, customerId, brand, model, serial);
    this.screenInches = screenInches;
}

Laptop.prototype = Object.create(Device.prototype);
Laptop.prototype.constructor = Laptop;

/**
 * Overrides the parent label and reuses it through an explicit `.call`.
 *
 * @returns {string}
 */
Laptop.prototype.label = function label() {
    return `${Device.prototype.label.call(this)} ${this.screenInches}"`;
};

File: rust/crates/atelier-core/src/models/device.rs
//! Device brought in for repair.

use serde::{Deserialize, Serialize};

/// A single serviceable unit belonging to a customer.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Device {
    pub id: u32,
    pub customer_id: u32,
    pub brand: String,
    pub model: String,
    pub serial: Option<String>,
}

impl Device {
    pub fn new(
        id: u32,
        customer_id: u32,
        brand: impl Into<String>,
        model: impl Into<String>,
    ) -> Self {
        Self {
            id,
            customer_id,
            brand: brand.into(),
            model: model.into(),
            serial: None,
        }
    }

    /// Human label such as `Framework 13 (SER-9)`.
    pub fn label(&self) -> String {
        match &self.serial {
            Some(serial) => format!("{} {} ({serial})", self.brand, self.model),
            None => format!("{} {}", self.brand, self.model),
        }
    }

    pub fn with_serial(mut self, serial: impl Into<String>) -> Self {
        self.serial = Some(serial.into());
        self
    }

    /// Devices without a serial cannot be warranty-claimed.
    pub fn is_warranty_eligible(&self) -> bool {
        self.serial.is_some()
    }
}
